### PR TITLE
rustdoc: when running a function-signature search, tweak the tab bar

### PR DESCRIPTION
**Pull request by notriddle**
# Before

In Names (7) / In Parameters (0) / In Return types (0)

# After

In Function Signature (7)

### CODE REVIEW

File: src/test/rustdoc-gui/search-tab-change-title-fn-sig.goml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// Checks that the search tab results work correctly with function signature syntax
+// First, try a search-by-name
+goto: file://|DOC_PATH|/test_docs/index.html
+write: (".search-input", "Foo")
+// Waiting for the search results to appear...
+wait-for: "#titles"
+assert-attribute: ("#titles > button:nth-of-type(1)", {"class": "selected"})
+assert-text: ("#titles > button:nth-of-type(1)", "In Names", CONTAINS)
```
Comment (GuillaumeGomez) [inline src/test/rustdoc-gui/search-tab-change-title-fn-sig.goml:8]:
```suggestion
assert-text: ("#titles > button:nth-of-type(1)", "In Names", STARTS_WITH)
```

Should work I think.

File: src/librustdoc/html/static/js/search.js [hunk 2]
```diff
@@ -1723,12 +1717,17 @@ window.initSearch = function(rawSearchIndex) {
...
             output += `<h3>Query parser error: "${results.query.error}".</h3>`;
+        } else if (results.query.foundElems <= 1 && results.query.returned.length === 0) {
+            output += `<div id="titles">` +
+                makeTabHeader(0, "In Names", ret_others[1]) +
+                makeTabHeader(1, "In Parameters", ret_in_args[1]) +
+                makeTabHeader(2, "In Return Types", ret_returned[1]) +
+                "</div>";
+        } else {
+            output += '<div id="titles">' +
+                makeTabHeader(0, "In Function Signature", ret_others[1]) +
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1728]: Should we differentiate between `a -> b` and `-> b`?

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1728]: "In function signature" and "In function returned value"? Also, does "function" include "method" as well?

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1728]: I guess you're right.

**Comment by GuillaumeGomez** [comment]
Considering it's a UI/UX change, I'd like to have @jsha and @Manishearth's opinion before going further. Personally I think it'd be a nice improvement.

**Comment by Manishearth** [comment]
I think it's a good call!

**Comment by GuillaumeGomez** [comment]
There is a JS error when the results are displayed:

Screenshot from 2022-05-02 20-29-56

The left/right arrows should be disabled in case there is only one column displayed, otherwise it looks like this (no "selected" header):

Screenshot from 2022-05-02 20-31-30

File: src/librustdoc/html/static/js/search.js [hunk 3]
```diff
@@ -1743,12 +1754,16 @@ window.initSearch = function(rawSearchIndex) {
...
         searchState.showResults(search);
         const elems = document.getElementById("titles").childNodes;
-        elems[0].onclick = function() { printTab(0); };
-        elems[1].onclick = function() { printTab(1); };
-        elems[2].onclick = function() { printTab(2); };
+        searchState.focusedByTab = [];
+        let i = 0;
+        for (const elem of elems) {
+            const j = i;
+            elem.onclick = function() { printTab(j); };
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1763]: I didn't know we could keep a variable alive this way. :o

**Comment by GuillaumeGomez** [comment]
I can confirm it's working locally as expected. :+1:

File: src/test/rustdoc-gui/search-tab-change-title-fn-sig.goml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+// Now try search-by-return
+goto: file://|DOC_PATH|/test_docs/index.html
+write: (".search-input", "-> String")
+// Waiting for the search results to appear...
+wait-for: "#titles"
+assert-attribute: ("#titles > button:nth-of-type(1)", {"class": "selected"})
+assert-text: ("#titles > button:nth-of-type(1)", "In Function Return Types", STARTS_WITH)
+// Use left-right keys
+press-key: "ArrowDown"
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/test/rustdoc-gui/search-tab-change-title-fn-sig.goml:28]: What I meant is that if the search input isn't focused, the down arrow won't have any effect, and therefore, the left and right arrow check will not do anything either.

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/test/rustdoc-gui/search-tab-change-title-fn-sig.goml:28]: Sorry if I wasn't clear. ^^'

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/test/rustdoc-gui/search-tab-change-title-fn-sig.goml:31]: Awesome, thanks!

**Comment by GuillaumeGomez** [comment]
Just one last weird thing:

Screenshot from 2022-05-03 21-41-03

It's a bit of strange rendering with doc aliases. Not a blocker though. You can try it with:

```rust
#[doc(alias = "->a")]
pub struct Foo;
```

Just confirming it's not an issue then let's go!

What do you think of this @Manishearth @jsha ? I personally think it's fine but asking just in case.

**Comment by jsha** [comment]
I think @notriddle's comment makes sense and I think the image @GuillaumeGomez showed is fine.

**Comment by GuillaumeGomez** [comment]
I said I was fine with it but thanks for the extra explanations. :) r=me once you fixed the conflict.

**Pull request merged**